### PR TITLE
Auth flow changes

### PR DESCRIPTION
**Pull request by avikalpg**
1. Logging in from any general page will keep you on the same page except the landing page - which sends you to u.tsx. Earlier, login would always send you to u.tsx
2. Pricing page: sends to u.tsx, now that login is currently handled in it. Also changed the applicability date to 31 Oct 2023
3. The login function in auth.ts now takes an option provider parameter that can directly let the user log in using a provider.
  i. Used this function in the docs to convert the first step into a link. Now the docs page can directly be shared.

### CODE REVIEW

File: pages/docs/index.tsx [hunk 1]
```diff
@@ -48,7 +48,14 @@ const Docs = ({ bitbucket_auth_url }: { bitbucket_auth_url: string }) => {
 			heading: "Github",
 			flag: true,
 			content: [
-				{ subHeading: "Sign up with github", article: "Sign in on Vibinex using GitHub" },
+				{
+					subHeading: "Sign up with github",
+					article: <span className="text-blue-500 cursor-pointer" onClick={() => login(
+						getAndSetAnonymousIdFromLocalStorage(),
```
Comment (tapishr) [inline pages/docs/index.tsx:54]: this is to help add event on login button, right?

File: pages/pricing.tsx [hunk 2]
```diff
@@ -63,7 +63,7 @@ const Pricing = () => {
 		return (<span className='font-money'>  {currency} <span className='text-4xl'>{price} </span ></span>);
 	}
 
-	const pricingStartDate = new Date(2023, 7, 31); // 31st August 2023
+	const pricingStartDate = new Date(2023, 9, 31); // 31st October 2023
```
Comment (tapishr) [inline pages/pricing.tsx:66]: This isnt used anywhere right? its a todo?

File: pages/u.tsx [hunk 3]
```diff
@@ -60,7 +61,9 @@ export const getServerSideProps: GetServerSideProps<ProfileProps> = async ({ req
 	if (!session) {
 		return {
 			redirect: {
-				destination: '/api/auth/signin',
+				destination: getURLWithParams('/api/auth/signin', {
+					callbackUrl: `${process.env.NEXTAUTH_URL}/u`
```
Comment (tapishr) [inline pages/u.tsx:65]: if !session then we are redirecting to u.tsx? Should we redirect to login page?

File: pages/pricing.tsx (hunk 2 above)
Comment (avikalpg) [reply pages/pricing.tsx:66]: This is used to show the "Applicable only from" text right below the heading.

File: pages/docs/index.tsx (hunk 1 above)
Comment (avikalpg) [reply pages/docs/index.tsx:54]: Yes.

File: pages/u.tsx (hunk 3 above)
Comment (avikalpg) [reply pages/u.tsx:65]: We are redirecting to `/api/auth/signin` with a parameter called `callbackUrl` which helps NextAuth understand that he has to send the user back to u.tsx once login is done.

**Review by tapishr — APPROVED** [review]

File: pages/pricing.tsx [hunk 4]
```diff
@@ -29,7 +29,7 @@ const pricingPlan = [
 		pricing: undefined, // will be populated by formula
 		features: ['Access of all features', 'Direct support through Slack', 'Free-of-cost setup assistance'],
 		buttonText: 'Start your 30 day trial',
-		link: '/api/auth/signin?callbackUrl=https%3A%2F%2Fvibinex.com%2Fpricing%2F' // temp. adding login page link, need to replace it with payment link
+		link: '/u'
```
Comment (MuskanPaliwal) [inline pages/pricing.tsx:32]: The comment will still be valid here right? Why did you remove it?

File: types/next-auth.d.ts [hunk 5]
```diff
@@ -11,4 +12,6 @@ declare module "next-auth" {
 			auth_info?: AuthInfo
 		}
 	}
+
+	type AuthProviderType = BuiltInProviderType | 'bitbucket'
```
Comment (MuskanPaliwal) [inline types/next-auth.d.ts:16]: so, bitbucket is not listed in the `BuiltInProviderType`?

**Review by MuskanPaliwal — APPROVED** [review]
Looks good to me, just a few queries.

File: pages/pricing.tsx (hunk 4 above)
Comment (avikalpg) [reply pages/pricing.tsx:32]: The button text says "Start your 30 day trial". We might not necessarily change the destination of the button click to a payment link. I feel that taking the user to the user profile (or login page and then the user profile page if he/she is not logged in) is actually a good way to handle this button click.

File: types/next-auth.d.ts (hunk 5 above)
Comment (avikalpg) [reply types/next-auth.d.ts:16]: Yes, it is not listed. Next-auth does not provide login with Bitbucket out of the box. We had to implement it.

Comment (MuskanPaliwal) [reply types/next-auth.d.ts:16]: Okayy

**Review by MuskanPaliwal — APPROVED** [review]
